File: templates/javascript/sliding-window.js
/**
 * Sliding Window Template - JavaScript
 * 
 * Use for: subarray sum, substring problems, fixed/variable window size
 * Time: O(n), Space: O(1)
 * 
 * Examples:
 * - Minimum Size Subarray Sum
 * - Longest Substring Without Repeating Characters
 * - Maximum Average Subarray I
 * - Subarray Sum Equals K
 */

function slidingWindow(nums, target) {
    let left = 0;
    let right = 0;
    let currentSum = 0;
    let result = 0;
    
    while (right < nums.length) {
        // Expand window
        currentSum += nums[right];
        
        // Shrink window when condition is met
        while (currentSum >= target) {
            // Update result
            result = Math.min(result, right - left + 1);
            
            // Remove left element
            currentSum -= nums[left];
            left++;
        }
        
        right++;
    }
    
    return result;
}

function fixedSizeWindow(nums, k) {
    let currentSum = 0;
    
    // Calculate sum of first window
    for (let i = 0; i < k; i++) {
        currentSum += nums[i];
    }
    
    let result = currentSum;
    
    // Slide window
    for (let i = k; i < nums.length; i++) {
        currentSum = currentSum - nums[i - k] + nums[i];
        result = Math.max(result, currentSum);
    }
    
    return result;
}

function variableSizeWindow(nums, target) {
    let left = 0;
    let currentSum = 0;
    let minLength = Infinity;
    
    for (let right = 0; right < nums.length; right++) {
        currentSum += nums[right];
        
        while (currentSum >= target) {
            minLength = Math.min(minLength, right - left + 1);
            currentSum -= nums[left];
            left++;
        }
    }
    
    return minLength === Infinity ? 0 : minLength;
}

File: templates/python/sliding-window.py
"""
Sliding Window Template - Python

Use for: subarray sum, substring problems, fixed/variable window size
Time: O(n), Space: O(1)

Examples:
- Minimum Size Subarray Sum
- Longest Substring Without Repeating Characters
- Maximum Average Subarray I
- Subarray Sum Equals K
"""

def sliding_window(nums, target):
    left = 0
    right = 0
    current_sum = 0
    result = 0
    
    while right < len(nums):
        # Expand window
        current_sum += nums[right]
        
        # Shrink window when condition is met
        while current_sum >= target:
            # Update result
            result = min(result, right - left + 1)
            
            # Remove left element
            current_sum -= nums[left]
            left += 1
        
        right += 1
    
    return result


def fixed_size_window(nums, k):
    current_sum = 0
    
    # Calculate sum of first window
    for i in range(k):
        current_sum += nums[i]
    
    result = current_sum
    
    # Slide window
    for i in range(k, len(nums)):
        current_sum = current_sum - nums[i - k] + nums[i]
        result = max(result, current_sum)
    
    return result


def variable_size_window(nums, target):
    left = 0
    current_sum = 0
    min_length = float('inf')
    
    for right in range(len(nums)):
        current_sum += nums[right]
        
        while current_sum >= target:
            min_length = min(min_length, right - left + 1)
            current_sum -= nums[left]
            left += 1
    
    return 0 if min_length == float('inf') else min_length


def longest_substring_without_repeating(s):
    char_set = set()
    left = 0
    max_length = 0
    
    for right in range(len(s)):
        while s[right] in char_set:
            char_set.remove(s[left])
            left += 1
        
        char_set.add(s[right])
        max_length = max(max_length, right - left + 1)
    
    return max_length
